src/main/java/com/abdullah/coding/challenge/services/BookingService.java

```diff
@@ -0,0 +1,359 @@
+import com.abdullah.coding.challenge.entities.Customer;
+import com.abdullah.coding.challenge.enums.BookingStatus;
+import com.abdullah.coding.challenge.enums.CustomerStatus;
+import com.abdullah.coding.challenge.repository.BookingRepository;
+import com.abdullah.coding.challenge.util.TimeUtil;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.sql.Timestamp;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Service
+public class BookingService {
+    @Autowired
+    private BookingRepository repository;
+    @Autowired
+    private CustomerService customerService;
+
+    @Autowired
+    private CabService cabService;
+
+    public Response<Booking> createBooking(BookingDTO bookingDTO) {
+        try {
+
+            if (bookingDTO.getCustomerId() <= 0) {
+                return new Response<>(HttpStatus.BAD_REQUEST, "Invalid Customer ID", null);
+            }
+
+            if (!TimeUtil.isValidTimeStamp(bookingDTO.getRequestTime())) {
+                return new Response<>(HttpStatus.BAD_REQUEST, "Invalid Date format", null);
+            }
+
+            Customer customer = customerService.getCustomer(bookingDTO.getCustomerId());
+            if (customer == null) {
+                return new Response<>(HttpStatus.NOT_FOUND, "No such Customer exists", null);
+            }
+
+            if (customer.getStatus().equals(CustomerStatus.INACTIVE.toString())) {
+                return new Response<>(HttpStatus.FORBIDDEN, "User blocked!", null);
+            }
+
+            Timestamp requestedTime = TimeUtil.getTimeStamp(bookingDTO.getRequestTime());
+            Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+            if (requestedTime == null || requestedTime.before(currentTime)) {
+                return new Response<>(HttpStatus.BAD_REQUEST, "Invalid Request Time, expected YYYY/MM/dd HH:mm", null);
+            }
+
+            Booking booking = new Booking();
+            booking.setStatus(String.valueOf(BookingStatus.REQUESTED));
+            booking.setBookingCode(UUID.randomUUID().toString());
+            booking.setVehicalType(bookingDTO.getVehicleType());
+            booking.setRequestTime(requestedTime);
+            booking.setCreationTime(currentTime);
+            booking.setCustomer(customer);
+
+            repository.save(booking);
+
+            return new Response<>(HttpStatus.CREATED, "Cab booking request sent!", booking);
+        } catch (Exception ex) {
+            return new Response<>(HttpStatus.INTERNAL_SERVER_ERROR, "Error processing the request!", null);
+        }
+    }
+
+    public Response<?> updateBooking(BookingDTO bookingDTO, String bookingCode) {
+        try {
+            Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+
+            if (bookingDTO.getCustomerId() <= 0) {
```

Comment:
1. The same validation code is used for booking creation. How can we avoid repetition?

Reply:
It's just one line of validation. You could wrap it in a function and jam your namespace. But I don't want to add many mini functions that just do one line validation.